File: app/pages/repos/_repo_id/models/_model_id/production_metrics/class_metrics/props.rs
use crate::{
	common::date_window::{DateWindow, DateWindowInterval},
	layouts::model_layout::ModelLayoutInfo,
};

#[derive(serde::Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct Props {
	pub id: String,
	pub class_metrics: Vec<ClassMetricsEntry>,
	pub date_window: DateWindow,
	pub date_window_interval: DateWindowInterval,
	pub classes: Vec<String>,
	pub overall: OverallClassMetrics,
	pub model_layout_info: ModelLayoutInfo,
	pub class: String,
}

#[derive(serde::Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct ClassMetricsEntry {
	pub class_name: String,
	pub intervals: Vec<IntervalEntry>,
}

#[derive(serde::Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct IntervalEntry {
	pub label: String,
	pub f1_score: TrainingProductionMetrics,
	pub precision: TrainingProductionMetrics,
	pub recall: TrainingProductionMetrics,
}

#[derive(serde::Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct OverallClassMetrics {
	pub class_metrics: Vec<OverallClassMetricsEntry>,
	pub label: String,
}

#[derive(serde::Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct OverallClassMetricsEntry {
	pub class_name: String,
	pub comparison: Comparison,
	pub confusion_matrix: ConfusionMatrix,
	pub f1_score: TrainingProductionMetrics,
	pub precision: TrainingProductionMetrics,
	pub recall: TrainingProductionMetrics,
}

#[derive(serde::Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct Comparison {
	pub false_negative_fraction: TrainingProductionMetrics,
	pub false_positive_fraction: TrainingProductionMetrics,
	pub true_positive_fraction: TrainingProductionMetrics,
	pub true_negative_fraction: TrainingProductionMetrics,
}

#[derive(serde::Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct ConfusionMatrix {
	pub false_negatives: Option<u64>,
	pub true_negatives: Option<u64>,
	pub true_positives: Option<u64>,
	pub false_positives: Option<u64>,
}

#[derive(serde::Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct TrainingProductionMetrics {
	pub production: Option<f32>,
	pub training: f32,
}

File: app/pages/repos/_repo_id/models/_model_id/production_metrics/class_metrics/props.ts
import { DateWindow, DateWindowInterval } from "common/time"
import { ModelLayoutInfo } from "layouts/model_layout"

export type Props = {
	class: string
	classMetrics: Array<{
		className: string
		intervals: Array<{
			f1Score: {
				production: number | null
				training: number
			}
			label: string
			precision: {
				production: number | null
				training: number
			}
			recall: {
				production: number | null
				training: number
			}
		}>
	}>
	classes: string[]

	dateWindow: DateWindow
	dateWindowInterval: DateWindowInterval
	id: string
	modelLayoutInfo: ModelLayoutInfo
	overall: {
		classMetrics: OverallClassMetrics[]
		label: string
	}

	title: string
}

export type OverallClassMetrics = {
	className: string
	comparison: {
		falseNegativeFraction: {
			production: number | null
			training: number
		}
		falsePositiveFraction: {
			production: number | null
			training: number
		}
		trueNegativeFraction: {
			production: number | null
			training: number
		}
		truePositiveFraction: {
			production: number | null
			training: number
		}
	}
	confusionMatrix: {
		falseNegatives: number | null
		falsePositives: number | null
		trueNegatives: number | null
		truePositives: number | null
	}
	f1Score: {
		production: number | null
		training: number
	}
	precision: {
		production: number | null
		training: number
	}
	recall: {
		production: number | null
		training: number
	}
}
